# **Waze Project**


# **PACE stages**

## **PACE: Analyze**

### **Task 2a. Imports and data loading**

In [1]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

In [2]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Summary information**

In [3]:
df.head()

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


Reviewing df.head() shows no missing values, but df.info() reveals 700 missing values in the 'label' column. Overall, the dataset has 14,999 rows and 13 columns, including 11 numeric columns and 2 of type object.

### **Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

In [5]:
# Isolate rows with null values
df_with_null = df[df.label.isna()]

# Display summary stats of rows with null values
df_with_null.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [6]:
# Isolate rows without null values
df_without_null = df[~df.label.isna()]

# Display summary stats of rows without null values
df_without_null.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

The two populations are similar in central tendencies, but there are some slightly differensses in mean values for total_sessions, driven_km_drives, and duration_minutes_drives.

### **Null values - device counts**

Next, let's check the two populations with respect to the `device` variable.

In [7]:
# Get count of null values by device

iphone_no_label = df[(df.device == 'iPhone') & (df.label.isna())].shape[0]
android_no_label = df[(df.device == 'Android') & (df.label.isna())].shape[0]

print(f"Number of iPhone users without a label: {iphone_no_label}")
print(f"Number of Android users without a label: {android_no_label}")

Number of iPhone users without a label: 447
Number of Android users without a label: 253


Now, of the rows with null values, let's calculate the percentage with each device&mdash;Android and iPhone. 

In [8]:
# Calculate % of iPhone nulls and Android nulls

# Calculate the percentage of each device (iPhone and Android)
device_counts = df_without_null.device.value_counts(normalize=True) * 100
print(device_counts)

device
iPhone     64.515001
Android    35.484999
Name: proportion, dtype: float64


Let's compare this to the device ratio in the full dataset

In [9]:
# Calculate % of iPhone users and Android users in full dataset
device_counts = df.device.value_counts(normalize=True) * 100
print(device_counts)

device
iPhone     64.484299
Android    35.515701
Name: proportion, dtype: float64


The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Let's examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [10]:
# Calculate counts of churned vs. retained
df.label.value_counts(normalize=True) * 100

label
retained    82.264494
churned     17.735506
Name: proportion, dtype: float64

This dataset contains 82% retained users and 18% churned users.

Next, let's compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that we don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [11]:
# Calculate median values of all columns for churned and retained users

churned_users = df[df['label'] == 'churned'].select_dtypes(include=['number']).median()
print(f"Churned users: \n{churned_users}\n")

retained_users = df[df['label'] == 'retained'].select_dtypes(include=['number']).median()
print(f"Retained users: \n{retained_users}")

Churned users: 
ID                         7477.500000
sessions                     59.000000
drives                       50.000000
total_sessions              164.339042
n_days_after_onboarding    1321.000000
total_navigations_fav1       84.500000
total_navigations_fav2       11.000000
driven_km_drives           3652.655666
duration_minutes_drives    1607.183785
activity_days                 8.000000
driving_days                  6.000000
dtype: float64

Retained users: 
ID                         7509.000000
sessions                     56.000000
drives                       47.000000
total_sessions              157.586756
n_days_after_onboarding    1843.000000
total_navigations_fav1       68.000000
total_navigations_fav2        9.000000
driven_km_drives           3464.684614
duration_minutes_drives    1458.046141
activity_days                17.000000
driving_days                 14.000000
dtype: float64


This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. 

To analyze the median kilometers per drive in the last month for both retained and churned users, we will first calculate the kilometers per drive by dividing the `driven_km_drives` column by the drives column. Once this ratio is obtained, we will group the data based on user status (churned or retained) and compute the median kilometers per drive for each group. This will provide insight into driving behavior differences between the two user segments.

In [12]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median, and isolate for km per drive
df[['label', 'km_per_drive']].groupby('label').median()

km_per_drive
label                 
churned      74.109416
retained     75.014702

The median retained user drove about one more kilometer per drive than the median churned user. How many kilometers per driving day was this?

In [64]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for km per driving day
df[['label', 'km_per_driving_day']].groupby('label').median()

km_per_driving_day
label                       
churned           697.541999
retained          289.549333

Now, let's calculate the median number of drives per driving day for each group.

In [13]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for drives per driving day
df[['label', 'drives_per_driving_day']].groupby('label').median()

drives_per_driving_day
label                           
churned                  10.0000
retained                  4.0625

The median user who churned drove 698 kilometers each day they drove last month, which is almost ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, let's examine whether there is an imbalance in how many users churned by device type.

The next step involves calculating the overall counts of each device type within the churned and retained groups. This will provide a breakdown of device usage across both segments, helping to better understand the distribution of device types among the different user groups.

In [14]:
# For each label, calculate the number of Android users and iPhone users
device_counts_by_label = df.groupby(['label', 'device']).size().unstack(fill_value=0)
print(device_counts_by_label)

device    Android  iPhone
label                    
churned       891    1645
retained     4183    7580


Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [15]:
# For each label, calculate the percentage of Android users and iPhone users
device_counts_by_label / device_counts_by_label.sum() * 100

device     Android     iPhone
label                        
churned   17.56011  17.831978
retained  82.43989  82.168022

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

## **PACE: Execute**

### **Conclusion**

**Questions:**

1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data?

   There are 700 missing values in the 'label' column. There is a potential pattern where missing values in the label column might be associated with specific devices, such as iPhones and Androids, as indicated by the analysis showing that 447 iPhone users and 253 Android users had missing labels.


2. What is a benefit of using the median value of a sample instead of the mean?

    A key benefit of using the median instead of the mean is that the median is less affected by outliers and extreme values. In the case of the driven_km_drives column, where the maximum value is 21,183 km, which is an unusually high value that could skew the mean, the median provides a more accurate representation of a typical user's driving distance. The median better reflects the central tendency of the data without being unduly influenced by extreme values.

3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?

    Yes, the investigation raised a few questions that could be worth exploring further:

    * Device-Specific Trends: Why are there more missing labels for iPhone and Android users? Is there a specific pattern or data collection issue tied to certain devices?

    * Churn and Retention Factors: What other factors beyond driving behavior (e.g., app usage frequency, time spent in the app) contribute to churn? Could there be an opportunity to gather additional data to understand this better?

    * Outlier Impact: Given the extreme outliers in certain columns (like driven_km_drives), would it be beneficial to apply data transformation or filtering methods to minimize their effect on analysis? Should outliers be handled differently across different user segments?

4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?

    In the dataset, 64.48% of the users are iPhone users, while 35.52% are Android users.

5. What were some distinguishing characteristics of users who churned vs. users who were retained?

   Churned users drove significantly more than retained users, with the median churned user covering 698 kilometers per driving day, about 240% more than retained users. Both groups are serious drivers, likely including many long-haul truckers. This suggests their needs may differ from typical drivers, and Waze could benefit from gathering more data on these high-mileage users to better tailor the app to their needs.

6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?

    There was no appreciable difference in churn rate between iPhone users and Android users. The ratio of iPhone to Android users is consistent across both the churned and retained groups, and these ratios align with the overall dataset.
